### ISSUE TITLE
Proposal: "function-default-params" scope

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint-scope#34]
From #33, eslint/eslint#9335.

As the step 27.a of [9.2.12 FunctionDeclarationInstantiation( func, argumentsList )](https://www.ecma-international.org/ecma-262/8.0/#sec-functiondeclarationinstantiation), `eslint-scope` should make new scope for the function body only if the parameters has any default value.

It's similar to `function-expression-name` scope which separates the variable of function expression names from function bodies. So I propose new `function-default-params` scope which separates the references in default parameters from function bodies.

1. If the current function has one or more default parameters in traversing, the referencer defines a `function-default-params` scope instead of `function` scope.
    - Note: the `function-default-params` scope has implicit `arguments` variable except if the function is arrow function expression.
    - Note: the `function-default-params` scope is the child scope of `function-expression-name` scope if the function is function expression.
2. The referencer evaluates the parameters.
3. The referencer defines `function` scope.
4. The referencer copies the all variables from the `function-default-params` scope to the `function` scope.
    - Note: those copies will prevent redeclaration such as `function f(a = 1) { let a }` since it's syntax errors.
    - Note: at least, `no-shadow` rule has to handle those copies as special.
    - Note: I think we should use shallow copy here because `Variable#defs` and `Variable#references` should be shared with both scopes but `Reference#resolved` property cannot have multiple variable objects. For example, in `function f(a, b = a) { g(a) }` case, the two references of `a` should be aggregated into one `Variable#references` array.
5. The referencer closes both `function-default-params` and `function` scopes at the end of the function.

**Comment by not-an-aardvark** [comment]
Is this still something we want to do?

**Comment by shannon** [comment]
I'm having a similar issue as https://github.com/estools/escope/issues/130 without this.

**Comment by mysticatea** [comment]
Oh, I had forgotten this issue. Yes, I hope to do. I added the memorandum of 7.

@shannon `eslint-scope` package has fixed in #33.

**Comment by kaicataldo** [comment]
@mysticatea Can we close this issue if it's fixed in #33?

**Comment by not-an-aardvark** [comment]
I don't think this issue has been fixed. If I'm understanding correctly, #33 fixed https://github.com/estools/escope/issues/130, which is a separate issue.

**Comment by mysticatea** [comment]
@kaicataldo No. The #33 was a superficial fix because I avoided breaking changes. And this is a follow‐up issue to make the correct scope structure.

But probably I should write an RFC.
